src/garage/tf/algos/te_npo.py

```diff
@@ -0,0 +1,1191 @@
+                                   inputs=flatten_inputs(
+                                       self._policy_opt_inputs),
+                                   constraint_name='mean_kl')
+
+        # Optimize inference distribution separately (supervised learning)
+        infer_loss, _ = self._build_inference_loss(infer_loss_inputs)
+        self.inference_optimizer.update_opt(loss=infer_loss,
+                                            target=self._inference,
+                                            inputs=flatten_inputs(
+                                                self._inference_opt_inputs))
+
+    def train(self, runner):
+        """Obtain samplers and start actual training for each epoch.
+
+        Args:
+            runner (LocalRunner): LocalRunner is passed to give algorithm
+                the access to runner.step_epochs(), which provides services
+                such as snapshotting and sampler control.
+
+        Returns:
+            float: The average return in last epoch cycle.
+
+        """
+        last_return = None
+
+        for _ in runner.step_epochs():
+            runner.step_path = runner.obtain_samples(runner.step_itr)
+            last_return = self.train_once(runner.step_itr, runner.step_path)
+            runner.step_itr += 1
+
+        return last_return
+
+    def train_once(self, itr, paths):
+        """Perform one step of policy optimization given one batch of samples.
+
+        Args:
+            itr (int): Iteration number.
+            paths (list[dict]): A list of collected paths.
+
+        Returns:
+            numpy.float64: Average return.
+
+        """
+        samples_data = self.paths_to_tensors(itr, paths)
+        logger.log('Optimizing policy...')
+        self.optimize_policy(itr, samples_data)
+        return samples_data['average_return']
+
+    def optimize_policy(self, itr, samples_data):
+        """Optimize policy.
+
+        Args:
+            itr (int): Iteration number.
+            samples_data (dict): Processed sample data.
+                See process_samples() for details.
+
+        """
+        del itr
+
+        policy_opt_input_values = self._policy_opt_input_values(samples_data)
+        inference_opt_input_values = self._inference_opt_input_values(
+            samples_data)
+
+        self._train_policy_and_encoder_networks(policy_opt_input_values)
+        self._train_inference_network(inference_opt_input_values)
+
+        paths = samples_data['paths']
+        self.evaluate(policy_opt_input_values, samples_data)
+        self.visualize_distribution()
+
+        logger.log('Fitting baseline...')
+        self._baseline.fit(paths)
+
+        self._old_policy.model.parameters = self.policy.model.parameters
+        self._old_policy.encoder.model.parameters = (
+            self.policy.encoder.model.parameters)
+        self._old_inference.model.parameters = self._inference.model.parameters
+
+    def paths_to_tensors(self, itr, paths):
```

Comment:
is there no way you can use `garage.np.paths_to_tensors` and then just add some extra info?